SpatialGDK/Source/SpatialGDK/Private/Interop/InitialOnlyFilter.cpp

```diff
@@ -0,0 +1,135 @@
+{
+	if (PendingInitialOnlyEntities.Num() == 0)
+	{
+		return;
+	}
+
+	TArray<Worker_Constraint> EntityConstraintArray;
+	TSet<Worker_EntityId_Key> EntitiesRequested(MoveTemp(PendingInitialOnlyEntities));
+
+	for (auto EntityId : EntitiesRequested)
+	{
+		UE_LOG(LogInitialOnlyFilter, Verbose, TEXT("Requested initial only data for entity %lld."), EntityId);
+
+		Worker_Constraint Constraints{};
+		Constraints.constraint_type = WORKER_CONSTRAINT_TYPE_ENTITY_ID;
+		Constraints.constraint.entity_id_constraint.entity_id = EntityId;
+
+		EntityConstraintArray.Add(Constraints);
+
+		InflightInitialOnlyEntities.Add(EntityId);
+	}
+
+	Worker_EntityQuery InitialOnlyQuery{};
+
+	InitialOnlyQuery.constraint.constraint_type = WORKER_CONSTRAINT_TYPE_OR;
+	InitialOnlyQuery.constraint.constraint.or_constraint.constraint_count = EntityConstraintArray.Num();
+	InitialOnlyQuery.constraint.constraint.or_constraint.constraints = EntityConstraintArray.GetData();
+	InitialOnlyQuery.snapshot_result_type_component_set_id_count = 1;
+	InitialOnlyQuery.snapshot_result_type_component_set_ids = &SpatialConstants::INITIAL_ONLY_COMPONENT_SET_ID;
+
+	const Worker_RequestId RequestID = NetDriver->Connection->SendEntityQueryRequest(&InitialOnlyQuery, SpatialGDK::RETRY_UNTIL_COMPLETE);
+	EntityQueryDelegate InitialOnlyQueryDelegate;
+	InitialOnlyQueryDelegate.BindRaw(this, &InitialOnlyFilter::HandleInitialOnlyResponse);
+
+	NetDriver->Receiver->AddEntityQueryDelegate(RequestID, InitialOnlyQueryDelegate);
+
+	InflightInitialOnlyRequests.Add(RequestID, { MoveTemp(EntitiesRequested) });
+}
+
+void InitialOnlyFilter::HandleInitialOnlyResponse(const Worker_EntityQueryResponseOp& Op)
+{
+	ClearRequest(Op.request_id);
+
+	if (Op.status_code != WORKER_STATUS_CODE_SUCCESS)
+	{
+		UE_LOG(LogInitialOnlyFilter, Error, TEXT("Failed to retrieve initial only data. Code: %d, %s"), Op.status_code,
+			   UTF8_TO_TCHAR(Op.message));
+		return;
+	}
+
+	for (uint32_t i = 0; i < Op.result_count; ++i)
+	{
+		const Worker_Entity* Entity = &Op.results[i];
+		const Worker_EntityId EntityId = Entity->entity_id;
+
+		// Ensure the entity we queried is still in view.
+		if (NetDriver->Connection->GetView().Find(EntityId) != nullptr)
+		{
+			UE_LOG(LogInitialOnlyFilter, Verbose, TEXT("Received initial only data for entity %lld."), EntityId);
+
+			// Extract and store the initial only data.
+			TArray<ComponentData>& ComponentDatas = RetrievedInitialOnlyData.FindOrAdd(EntityId);
+			for (uint32_t j = 0; j < Entity->component_count; ++j)
+			{
+				const Worker_ComponentData* ComponentData = &Entity->components[j];
+				ComponentDatas.Emplace(ComponentData::CreateCopy(ComponentData->schema_type, ComponentData->component_id));
+			}
+
+			NetDriver->Connection->GetCoordinator().RefreshEntityCompleteness(Entity->entity_id);
+		}
+	}
+}
+
+const TArray<SpatialGDK::ComponentData>& InitialOnlyFilter::GetInitialOnlyData(Worker_EntityId EntityId) const
+{
+	return *RetrievedInitialOnlyData.Find(EntityId);
```

Comment:
Might be worth adding a check here, we ask users of this class to check existence with HasInitialOnlyData first but it may not be obvious in the future. Perhaps it could be combined to return a const pointer instead, which could be null

Reply:
(And omits an additional hash lookup)

Reply:
Changed to return pointer